src/datasources/cache/redis.cache.service.ts

```diff
@@ -42,11 +43,26 @@ export class RedisCacheService implements ICacheService, OnModuleDestroy {
 
   /**
    * Closes the connection to Redis when the module associated with this service
-   * is destroyed.
+   * is destroyed. This tries to gracefully close the connection. If the Redis
+   * instance is not responding it invokes {@link forceQuit}.
    */
   async onModuleDestroy(): Promise<void> {
     this.logger.verbose('Closing Redis connection');
+    const forceQuitTimeout = setTimeout(
+      this.forceQuit.bind(this),
+      this.quitTimeoutInSeconds * 1000,
+    );
     await this.client.quit();
+    clearTimeout(forceQuitTimeout);
+    this.logger.verbose('Redis connection closed');
+  }
+
+  /**
+   * Forces the closing of the Redis connection associated with this service.
+   */
+  private forceQuit() {
+    this.logger.verbose('Forcing Redis connection close');
+    this.client.disconnect();
```

Comment:
`disconnect` returns a `Promise` shouldn't we make `forceQuit` async? How does that play with the `bind()` in `setTimeout`?

Reply:
Yes, maybe it's more correct to define it as `async` and wait for the Promise to be resolved. Solved [here.](https://github.com/5afe/safe-client-gateway-nest/pull/83/commits/968af6e807c8c9e3da58ec472ae40d3512573e83) Thanks.

`bind()` creates a new function but doing a closure over `this` (i.e. save the context of the caller). Otherwise `setTimeout` callback (`forceQuit` function) won't have access to the context when fired. This is necessary to still having access to `this.client` when invoking `disconnect`.

Reply:
Do you know about: https://nodejs.org/api/timers.html#timers-promises-api

I did a quick mockup with the following:

```typescript
import { setTimeout } from 'timers/promises';

...

async onModuleDestroy(): Promise<void> {
  const abortController = new AbortController();
  this.logger.verbose('Closing Redis connection');

  (async () => {
    await setTimeout(this.quitTimeoutInSeconds * 1000, this.forceQuit(), {
      signal: abortController.signal,
    });
  })();

  await this.client.quit();
  abortController.abort();
  this.logger.verbose('Redis connection closed');
}
```